### PR TITLE
feat: database update attribute endpoints

### PR DESCRIPTION
**Pull request by TorstenDittmann**
## What does this PR do?

- [x] adds `PATCH` endpoints for each attribute 
- [x] adds references to each endpoint for documentation
- [x] adds 2 more exceptions
  - `ATTRIBUTE_TYPE_INVALID`
  - `ATTRIBUTE_FILTER_INVALID`
- [] add tests
- [] update documentation
- [] update contribution guide

## Test Plan

- [] test against success
- [] test against failures

## Related PRs and Issues

- #2720 

## Checklist

- [x] Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?
- [] If the PR includes a change to an API's metadata (desc, label, params, etc.), does it also include updated API specs and example docs?

### CODE REVIEW

File: tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php [hunk 1]
```diff
@@ -484,16 +484,6 @@ public function testUpdateFunction($data): array
...
-        $function = $this->client->call(Client::METHOD_PUT, '/functions/' . $data['functionId'], array_merge([
-            'content-type' => 'application/json',
-            'x-appwrite-project' => $this->getProject()['$id'],
-        ], $this->getHeaders()), [
-            'name' => 'Test Failure',
-            'execute' => [ 'not-valid-permission' ]
-        ]);
-
-        $this->assertEquals($function['headers']['status-code'], 400);
-
```
Comment (christyjacob4) [inline tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php:496]: Any reason we removed this test?

File: src/Appwrite/GraphQL/Types/Mapper.php [hunk 2]
```diff
@@ -109,9 +109,6 @@ public static function route(
                     'type' => $parameterType,
                     'description' => $parameter['description'],
                 ];
-                if ($parameter['optional']) {
-                    $params[$name]['defaultValue'] = $parameter['default'];
-                }
```
Comment (christyjacob4) [inline src/Appwrite/GraphQL/Types/Mapper.php:114]: why did we remove this ?

File: app/controllers/api/databases.php [hunk 3]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateStringAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-string-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1639]:
```suggestion
    ->label('sdk.response.code', Response:: STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 4]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateEmailAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-email-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1680]:
```suggestion
    ->label('sdk.response.code', Response:: STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 5]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateEnumAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-enum-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1722]:
```suggestion
    ->label('sdk.response.code', Response:: STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 6]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateIpAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-ip-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1766]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 7]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateUrlAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-url-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1808]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 8]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateIntegerAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-integer-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1849]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 9]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateFloatAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-float-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1894]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 10]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateBooleantAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-boolean-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1939]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

File: app/controllers/api/databases.php [hunk 11]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('event', 'databases.[databaseId].collections.[collectionId].attributes.[attributeId].update')
+    ->label('audits.event', 'attribute.update')
+    ->label('audits.resource', 'database/{request.databaseId}/collection/{request.collectionId}')
+    ->label('usage.metric', 'collections.{scope}.requests.update')
+    ->label('usage.params', ['databaseId:{request.databaseId}'])
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateDatetimeAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-datetime-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
```
Comment (christyjacob4) [inline app/controllers/api/databases.php:1980]:
```suggestion
    ->label('sdk.response.code', Response::STATUS_CODE_ACCEPTED)
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Please address the comments

File: tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php (hunk 1 above)
Comment (TorstenDittmann) [reply tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php:496]: no need to test against failure in webhook tests 👍🏻

File: app/controllers/api/databases.php [hunk 12]
```diff
@@ -1483,7 +2048,7 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
         }
 
         // Only update status if removing available attribute
-        if ($attribute->getAttribute('status' === 'available')) {
+        if ($attribute->getAttribute('status') === 'available') {
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:2051]: yes 🙃

File: src/Appwrite/GraphQL/Types/Mapper.php (hunk 2 above)
Comment (TorstenDittmann) [reply src/Appwrite/GraphQL/Types/Mapper.php:114]: this is handled by utopia framework now 👍🏻

File: app/controllers/api/databases.php [hunk 13]
```diff
@@ -1440,6 +1623,388 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateDatetimeAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-datetime-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.model', Response::MODEL_ATTRIBUTE_DATETIME)
+    ->param('databaseId', '', new UID(), 'Database ID.')
+    ->param('collectionId', '', new UID(), 'Collection ID. You can create a new collection using the Database service [server integration](https://appwrite.io/docs/server/databases#databasesCreateCollection).')
+    ->param('key', '', new Key(), 'Attribute Key.')
+    ->param('required', null, new Boolean(), 'Is attribute required?')
+    ->param('default', null, new Nullable(new DatetimeValidator()), 'Default value for attribute when not provided. Cannot be set when attribute is required.')
```
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1986]: right now we only allow updating metadata, nothing that actually would need changes in the documents itself.

If I make this param optional, I don't know if it defaulted to `null` or `null` is sent 👍🏻

File: app/controllers/api/databases.php (hunk 4 above)
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1680]: `202` doesn't apply, because it is not handled by any worker

> The HyperText Transfer Protocol (HTTP) 202 Accepted response status code indicates that the request has been accepted for processing, but the processing has not been completed; in fact, processing may not have started yet.

File: tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php (hunk 1 above)
Comment (christyjacob4) [reply tests/e2e/Services/Webhooks/WebhooksCustomServerTest.php:496]: okay 👍

File: app/controllers/api/databases.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (christyjacob4) [reply app/controllers/api/databases.php:2051]: Yikes 😬

File: app/controllers/api/databases.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (christyjacob4) [reply app/controllers/api/databases.php:1986]: We should also update the `createAttribute` endpoints and make the default there Nullable as well. 

https://github.com/appwrite/appwrite/blob/bfd23665cb5ea0588dbad6d8f8303020096f11d4/app/controllers/api/databases.php#LL1057C22-L1057C140

File: src/Appwrite/GraphQL/Types/Mapper.php (hunk 2 above)
Comment (christyjacob4) [reply src/Appwrite/GraphQL/Types/Mapper.php:114]: okay

File: app/controllers/api/databases.php (hunk 4 above)
Comment (christyjacob4) [reply app/controllers/api/databases.php:1680]: in that case we must update the return status codes to be 200 as well 

Screenshot 2023-03-02 at 12 04 34 PM

File: app/controllers/api/databases.php (hunk 13 above)
[the remark this answers is not in the document]
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1986]: > We should also update the `createAttribute` endpoints and make the default there Nullable as well.
> 
> https://github.com/appwrite/appwrite/blob/bfd23665cb5ea0588dbad6d8f8303020096f11d4/app/controllers/api/databases.php#LL1057C22-L1057C140

there we don't necessarily need a difference 🤔 we could save that breaking change

File: app/controllers/api/databases.php (hunk 4 above)
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1680]: oopsie

**Review by gewenyu99 — APPROVED** [review]
Docs look good.

**Comment by lohanidamodar** [comment]
@TorstenDittmann can we also update one of the mock endpont  for tests (required nullable value)?

File: app/controllers/api/databases.php [hunk 14]
```diff
@@ -147,6 +148,163 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+            }
+
+            $options = [
+                'min' => $min,
+                'max' => $max
+            ];
+            $attribute->setAttribute('formatOptions', $options);
+
+            break;
+        case APP_DATABASE_ATTRIBUTE_FLOAT_RANGE:
```
Comment (abnegate) [inline app/controllers/api/databases.php:234]: Are we able to combine these two cases?

File: app/controllers/api/databases.php [hunk 15]
```diff
@@ -147,6 +148,163 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+            }
+
+            //TODO: before merging - this iteration is kinda hard to follow because of the $size variable?
+            $size = 0;
+            foreach ($elements as $element) {
+                $length = \strlen($element);
+                if ($length === 0) {
+                    throw new Exception(Exception::ATTRIBUTE_VALUE_INVALID, 'Each enum element must not be empty');
+                }
+                $size = ($length > $size) ? $length : $size;
```
Comment (abnegate) [inline app/controllers/api/databases.php:272]: Do we actually need the `$size`? We are assigning it but never reading

File: app/controllers/api/databases.php [hunk 16]
```diff
@@ -1440,6 +1599,394 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
...
+    ->label('sdk.auth', [APP_AUTH_TYPE_KEY])
+    ->label('sdk.namespace', 'databases')
+    ->label('sdk.method', 'updateStringAttribute')
+    ->label('sdk.description', '/docs/references/databases/update-string-attribute.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.model', Response::MODEL_ATTRIBUTE_STRING)
+    ->param('databaseId', '', new UID(), 'Database ID.')
+    ->param('collectionId', '', new UID(), 'Collection ID. You can create a new collection using the Database service [server integration](https://appwrite.io/docs/server/databases#databasesCreateCollection).')
+    ->param('key', '', new Key(), 'Attribute Key.')
+    ->param('required', null, new Boolean(), 'Is attribute required?')
```
Comment (abnegate) [inline app/controllers/api/databases.php:1620]: If this is a required param but not `Nullable`, how can the default be `null`?

Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1620]: The default is never used here, because it is required.

**Review by abnegate — APPROVED** [review]

File: app/controllers/api/databases.php [hunk 17]
```diff
@@ -1440,6 +1579,394 @@ function createAttribute(string $databaseId, string $collectionId, Document $att
         $response->dynamic($attribute, $model);
     });
 
+App::patch('/v1/databases/:databaseId/collections/:collectionId/attributes/:key/string')
```
Comment (abnegate) [inline app/controllers/api/databases.php:1582]: Do you think it would make more sense if the paths were `/type/key` instead of `/key/type`

E.g.
`v1/databases/marvel/collections/movies/attributes/string/name` instead of
`v1/databases/marvel/collections/movies/attributes/name/string`

**Review by christyjacob4 — APPROVED** [review]

File: app/controllers/api/databases.php (hunk 17 above)
Comment (TorstenDittmann) [reply app/controllers/api/databases.php:1582]: I think you're right 🤔
